src/Extensions/DoctrineResultDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+declare(strict_types=1);
+
+namespace staabm\PHPStanDba\Extensions;
+
+use Doctrine\DBAL\Result;
+use PhpParser\Node\Expr\MethodCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\ShouldNotHappenException;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\Constant\ConstantArrayType;
+use PHPStan\Type\Constant\ConstantArrayTypeBuilder;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\DynamicMethodReturnTypeExtension;
+use PHPStan\Type\Generic\GenericObjectType;
+use PHPStan\Type\IntegerType;
+use PHPStan\Type\Type;
+use staabm\PHPStanDba\QueryReflection\QueryReflector;
+
+final class DoctrineResultDynamicReturnTypeExtension implements DynamicMethodReturnTypeExtension
+{
+    public function getClass(): string
+    {
+        return Result::class;
+    }
+
+    public function isMethodSupported(MethodReflection $methodReflection): bool
+    {
+        return \in_array(strtolower($methodReflection->getName()), ['columncount', 'fetchnumeric', 'fetchassociative', 'fetchallnumeric', 'fetchallassociative'], true);
+    }
+
+    public function getTypeFromMethodCall(MethodReflection $methodReflection, MethodCall $methodCall, Scope $scope): Type
+    {
+        $defaultReturn = ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
+
+        $resultType = $scope->getType($methodCall->var);
+
+        if (!$resultType instanceof GenericObjectType) {
+            return $defaultReturn;
+        }
+
+        $genericTypes = $resultType->getTypes();
+
+        if (1 !== \count($genericTypes)) {
+            return $defaultReturn;
+        }
+
+        $resultRowType = $genericTypes[0];
+
+        switch (strtolower($methodReflection->getName())) {
+            case 'columncount':
+                if ($resultRowType instanceof ConstantArrayType) {
+                    $columnCount = \count($resultRowType->getKeyTypes()) / 2;
```

Comment:
I'm not sure if you can simply divide here, it may depend on the FETCH_TYPE used when executing the statement? As in if all keys are numeric or all keys are associative, then you shouldn't divide by 2? Unless I misunderstand what is going on here.. :)

Reply:
Ah I guess in a static analysis context this is not true as you always have all columns listed. Never mind.